Change testing setup to allow javalib testing against the JVM

nezosb opened the pull request:
Added the ability to run java.* `unit-tests` and `unit-tests-ext` in a JVM. To run `unit-tests` in JVM use testsJVM/test and to run `unit-tests-ext` use testsExtJVM/test. Similarly to Scala.js both modules were divided into `jvm`, `native` and `shared`. java.* tests were put in `shared`, other tests were left in `native`. To avoid JVM SecurityErrors, packages in `shared` were renamed from java.* to javalib.* and subsequently, files that relied on values in the java.* package scope (or depended on Scala Native modules) were left untouched in `native`.

Since multiple tests ended up not passing in JVM, blacklisting system was also added, similar to the one used in the recently added partests. I imagine, as javalib module will be developed further, fewer tests will have to be blacklisted. Some inconsistencies there seem trivial, like incorrect exceptions being thrown, though few are much more complex, so I decided that fixing all of those was beyond the scope of this PR, although I would like to perhaps try to fix those at a later date.

This PR should resolve #2124.

nezosb commented:
Hmm, ProcessTest (for the SN tests, not newly added JVM) keeps timing out in the CI with quite the frequency now... Perhaps adding additional tasks (testsJVM/test and testsExtJVM/test) to `test-runtime` only escalated the problem. I'm not sure if there is a point to restarting the CI like this, so I will remove testsJVM/test and testsExtJVM/test previously added in this PR from `test-runtime`. Maybe that will be for the better long term wise.

nezosb commented:
That did not help at all. Oh well, I'll look more into it tomorrow.

ekrich commented:
@nezosb Thank-you for this contribution, we really need this feature. Some of the main contributors/reviewers are out on vacation so it may take a couple of weeks so please be patient.

WojciechMazur reviewed (changes requested):
Good work, look well in overall, I have only some minor/styling fixes. 
I've rerun CI, now ProcessTest has failed in two tasks. Increasing timeout would probably not help us to fix that. Also, I've not observed similar failures in other PRs which suggest that's not related to CI env. I had had two guesses there might be some problem with not closed IO streams used by the created process, I had some similar issues with Windows recently, but it's probably not directly related. 
We definitely need to place this PR on hold before fixing the underlying issue.

nezosb commented on unit-tests/jvm/src/test/scala/utils/Platform.scala:
```diff
@@ -0,0 +1,26 @@
+
+  final val executingInJVM = true
+
+  final val executingInScalaJS = false
```
It is used `ArraysTest.scala`, so I imagine source compliance was the reason. An alternative would be to remove the checks in ArraysTest, but I don't really have any strong opinions about this. Perhaps it will be slightly easier in the long run if we leave it as it is.

WojciechMazur replied:
Let's keep it then

nezosb commented:
As per @WojciechMazur's request, I have added more tests to `shared`, mainly of the scalalib. I have also split `ProcessTest` into tests which can be run in JVM (`ProcessTest.scala`), and the one who cannot (`ProcessInheritTest.scala`). Along with those changes, more tests had to be added to the blacklist.

WojciechMazur commented:
It looks good now, I've also found few other tests that should be moved to shared: `java.lang.SystemTest`, `scala.AsInstanceOfTest`, `scala.HashCodeTest`, none of them should cause any problems when moved.  
Also probably `re2-eshastive.txt.gz` should be moved from shared to native resources.

nezosb reviewed (commented):
I have moved the aforementioned tests to shared. SystemTest I had to split similarly to ProcessTest. After running that on JVM, I discovered that envVars were not being set there, so I adjusted that in build.sbt. For the other two tests to compile, I had to add BuildInfo to testsJVM.
